source/Scope.cpp

```diff
@@ -4,18 +4,32 @@
 namespace slang {
 
 static const Scope emptyScope;
-const Scope* const Scope::Empty = &emptyScope;
+const Scope *const Scope::Empty = &emptyScope;
 
-void Scope::add(const Symbol* symbol) {
-    ASSERT(symbol);
-    table.emplace(symbol->name, symbol);
+bool Scope::add(const Symbol *symbol) {
+        ASSERT(symbol);
+    list.emplace_back(symbol);
+    return table.emplace(symbol->name, symbol).second;
 }
 
-const Symbol* Scope::lookup(StringRef name) const {
-    auto it = table.find(name);
-    if (it == table.end())
-        return nullptr;
-    return it->second;
+const Symbol *Scope::lookup(StringRef name, bool local) const {
+    for (auto scope = this; scope != Scope::Empty; scope = scope->parent()) {
+        auto it = table.find(name);
+        if (it != table.end()) return it->second;
+        if (local) return nullptr;
+    }
+    return nullptr;
 }
 
-}
\ No newline at end of file
+const Symbol* Scope::getNth(const SymbolKind& kind, size_t index) const {
```

Comment:
Is there some sort of guarantee of ordering so we can actually know what this will do? I understand the normal lookups, I don't really understand the idea behind this.